From 667758996a2301b74beaacb198f0e30613aaf923 Mon Sep 17 00:00:00 2001
From: Simo Piiroinen <simo.piiroinen@jolla.com>
Date: Fri, 6 Nov 2020 10:13:35 +0200
Subject: [PATCH] Allow changing "protocol" list after initial set

Firejail uses set-once logic for "protocol" list. This makes it
impossible to accumulate list of allowed protocols from multiple
include files.

Use profile_list_augment() for maintaining list of protocols. This
implicitly means protocols can be added/removed via any number of
command line options / profile configuration files.

Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>
---
 src/firejail/main.c    | 12 +++---------
 src/firejail/profile.c | 12 +++---------
 2 files changed, 6 insertions(+), 18 deletions(-)

diff --git a/src/firejail/main.c b/src/firejail/main.c
index 5cc2d412..d0ed1871 100644
--- a/src/firejail/main.c
+++ b/src/firejail/main.c
@@ -1269,15 +1269,9 @@ int main(int argc, char **argv, char **envp) {
 #endif
 		else if (strncmp(argv[i], "--protocol=", 11) == 0) {
 			if (checkcfg(CFG_SECCOMP)) {
-				if (cfg.protocol) {
-					fwarning("two protocol lists are present, \"%s\" will be installed\n", cfg.protocol);
-				}
-				else {
-					// store list
-					cfg.protocol = strdup(argv[i] + 11);
-					if (!cfg.protocol)
-						errExit("strdup");
-				}
+				const char *add = argv[i] + 11; // Just OMG...
+				profile_list_augment(&cfg.protocol, add);
+				fwarning("[main] combined protocol list: \"%s\"\n", cfg.protocol);
 			}
 			else
 				exit_err_feature("seccomp");
diff --git a/src/firejail/profile.c b/src/firejail/profile.c
index c04ab4c5..724758a2 100644
--- a/src/firejail/profile.c
+++ b/src/firejail/profile.c
@@ -892,15 +892,9 @@ int profile_check_line(char *ptr, int lineno, const char *fname) {
 
 	if (strncmp(ptr, "protocol ", 9) == 0) {
 		if (checkcfg(CFG_SECCOMP)) {
-			if (cfg.protocol) {
-				fwarning("two protocol lists are present, \"%s\" will be installed\n", cfg.protocol);
-				return 0;
-			}
-
-			// store list
-			cfg.protocol = strdup(ptr + 9);
-			if (!cfg.protocol)
-				errExit("strdup");
+			const char *add = ptr+9; // Just OMG...
+			profile_list_augment(&cfg.protocol, add);
+			fwarning("[profile] combined protocol list: \"%s\"\n", cfg.protocol);
 		}
 		else
 			warning_feature_disabled("seccomp");
-- 
2.17.1

